WIP - Implement UUID-friendly Vault Economy provider

mdcfe opened the pull request:
***Work-in-progress** - do not merge.*

This PR aims to implement a new `net.milkbowl.vault.economy.Economy` provider for EssentialsX that upgrades requests with usernames to requests with `OfflinePlayer`s, and makes use of `IUser` directly to manipulate the economy rather than `com.earth2me.essentials.api.Economy`.

This is necessary as the legacy `Economy_Essentials` provider included in Vault downgrades `OfflinePlayer`-based method calls to username-based calls, for whatever reason, which causes issues such as #913 when `OfflinePlayer`s with null names throw errors in Essentials's `Economy` API. (Vault's decision to downgrade requests also led to #1642, which was solved in LuckPerms with lucko/LuckPerms@5a34fa1.)

This implementation also lets us freely extend EssentialsEco beyond what was exposed by Vault's legacy provider, such as adding proper bank account support to the Vault provider without plugins needing to hook directly into EssentialsX.

In addition, this PR includes a method that deregisters the original `Economy_Essentials` provider, which means that our provider can replace the legacy one, while still allowing any third party economy plugins to take priority if installed.

---

Something to consider is whether we should include bank account support or not - I'd have to look into whether this would require significant changes or not. It's worth noting that EssentialsX's current support for NPCs seems to only exist via `com.earth2me.essentials.api.Economy` (excluding plugins like Citizens).

mdcfe committed:
Add Vault hook manager

Responsible for interactions with service manager to register and deregister hooks as necessary.

mdcfe committed:
Plugins are not providers

Deregister the specific EssEco provider, not just every single Vault provider ever because that's utterly ridiculous and stupid and why would anyone do that *glares at self*

mdcfe committed:
Add incomplete economy provider

mdcfe committed:
Implement more economy methods

mdcfe committed:
Make Vault hook features configurable

mdcfe committed:
Rename provider management class and methods

Now with 10% more accurate naming!

caojohnny commented on Essentials/src/config.yml:
```diff
@@ -612,6 +612,15 @@ minimum-pay-amount: 0.001
+
+# Whether to disable Vault's legacy EssentialsEco hook or not.
+# This may be necessary to avoid conflicts with EssentialsX.
+# This will only work if use-new-vault-hook is enabled!
```
I'm curious as to what would happen if both were set to `false`. Would you disable the plugin and print out an error then? Or will you simply override what is configured and load the new vault hook?

mdcfe replied:
If both are set to false, this does nothing - Vault's built in Essentials hook will be used instead.

caojohnny replied:
I see.

md678685 committed:
Combine Vault provider settings

There isn't any need to separate these realistically - if ours is active
then the Vault one won't be.

md678685 committed:
Implement bank not supported messages

md678685 committed:
Implement depositPlayer, unsupported bank responses

Totally didn't forget to do the bank messages in the last commit, nope
